From b7cdb25e8245281877c0ecd7ec79128460aedfbe Mon Sep 17 00:00:00 2001
From: Martin Kosek <mkosek@redhat.com>
Date: Fri, 13 Jun 2014 12:07:08 +0200
Subject: [PATCH 107/116] Administrative password change does not respect
 password policy

When Directory Manager or a PassSync agent is changing a password,
it is not being expired, but standard expiration time should apply.
However, default expiration time was always applied (90 days)
even though administrator may have a custom policy for the user.

https://fedorahosted.org/freeipa/ticket/3968
---
 .../ipa-pwd-extop/ipapwd_common.c                  | 47 +++++++++++++++-------
 1 file changed, 32 insertions(+), 15 deletions(-)

diff --git a/daemons/ipa-slapi-plugins/ipa-pwd-extop/ipapwd_common.c b/daemons/ipa-slapi-plugins/ipa-pwd-extop/ipapwd_common.c
index cac8bf45f3770e102c49735335066b7bc761dba2..e57a288abf1947d28d2494cf09e7c91d3103e293 100644
--- a/daemons/ipa-slapi-plugins/ipa-pwd-extop/ipapwd_common.c
+++ b/daemons/ipa-slapi-plugins/ipa-pwd-extop/ipapwd_common.c
@@ -625,6 +625,7 @@ done:
 int ipapwd_CheckPolicy(struct ipapwd_data *data)
 {
     struct ipapwd_policy pol = {0};
+    struct ipapwd_policy tmppol = {0};
     time_t acct_expiration;
     time_t pwd_expiration;
     time_t last_pwd_change;
@@ -635,22 +636,38 @@ int ipapwd_CheckPolicy(struct ipapwd_data *data)
     pol.max_pwd_life = IPAPWD_DEFAULT_PWDLIFE;
     pol.min_pwd_length = IPAPWD_DEFAULT_MINLEN;
 
-    if (data->changetype != IPA_CHANGETYPE_NORMAL) {
-        /* We must skip policy checks (Admin change) but
-         * force a password change on the next login.
-         * But not if Directory Manager */
-        if (data->changetype == IPA_CHANGETYPE_ADMIN) {
+    switch(data->changetype) {
+        case IPA_CHANGETYPE_ADMIN:
+            /* We must skip policy checks (Admin change) but force a password
+             * change on the next login.
+             */
             data->expireTime = data->timeNow;
-        }
-
-        /* do not load policies */
-    } else {
-
-        /* find the entry with the password policy */
-        ret = ipapwd_getPolicy(data->dn, data->target, &pol);
-        if (ret) {
-            LOG_TRACE("No password policy, use defaults");
-        }
+            break;
+        case IPA_CHANGETYPE_NORMAL:
+            /* Find the entry with the password policy */
+            ret = ipapwd_getPolicy(data->dn, data->target, &pol);
+            if (ret) {
+                LOG_TRACE("No password policy, use defaults");
+            }
+            break;
+        case IPA_CHANGETYPE_DSMGR:
+            /* PassSync agents and Directory Manager can administratively
+             * change the password without expiring it.
+             *
+             * Find password policy for the entry to properly set expiration.
+             * Do not store it in resulting policy to avoid aplying password
+             * quality checks on administratively set passwords
+             */
+            ret = ipapwd_getPolicy(data->dn, data->target, &tmppol);
+            if (ret) {
+                LOG_TRACE("No password policy, use defaults");
+            } else {
+                pol.max_pwd_life = tmppol.max_pwd_life;
+            }
+            break;
+        default:
+            LOG_TRACE("Unknown password change type, use defaults");
+            break;
     }
 
     tmpstr = slapi_entry_attr_get_charptr(data->target,
-- 
1.9.3

